Opening explorer

Mauritz8 opened the pull request:
Only supports the master database for now. Thought it would be a good idea to create a pr before I go any further.
demo: [Screencast from 2024-07-06 22:35:13.webm](https://github.com/lichess-org/mobile/assets/67279312/f559b5e0-786d-4cb2-a492-5b1d9b8d3dea)

HaonRekcef commented:
Hi Mauritz
Thanks, looks promising.
I was working on the opening explorer feature a few months back, but had to stop due to other personal obligations. Unfortunately, I don't expect this to change in the next few months :(. I have published my progress here, so maybe you can take some inspiration:
https://github.com/HaonRekcef/lichess-mobile/tree/opening-explorer
Here's a screen recording of how it looked:

https://github.com/lichess-org/mobile/assets/78898963/2bc445d9-ea35-409a-89a7-fdf85c02f1ee

Note that the code and the design are not polished. If I remember correctly, the communication with the server is not done well, and there are some other things not implemented. 

Some feedback and points to discuss:
1. I am not sure if there is space for the opening explorer in the analysis board, even though I loved it there in the old app. If there is some notation and the engine is turned on, it won’t fit. 
2. I experimented with different types of Flutter tables a lot, and I think the data table widget takes up too much space vertically. I managed to get a working version with the Table widget that uses less space while still having the ink behavior of the data table widget (highlighting the whole row).  See here: https://github.com/HaonRekcef/lichess-mobile/blob/e65893273b07b3bfb2512dc084a3ad75e060d44f/lib/src/view/explorer/opening_explorer.dart#L106C23-L224C23 
3. In my opinion, an opening explorer without a way to view games is incomplete.

Mauritz8 commented:
Hi! Thanks for taking the time to share your thoughts and give feedback, I appreciate it more than you know.

To address your points:
1. I agree that it should be a separate screen, right now you have to scroll a lot and it's hard to get a good overview.
2. I agree on this one as well, I think the less scrolling needed the better. I'll check out the way you used the table widget.
3. Definitely, I will add it.

On a side note, I tried to get each color block in the percentage graph to have widths according to their percentage, but I had a lot of problems with the table changing size depending on the data and sometimes going of screen, and I couldn't really figure out why. Do you remember if you had any similar problems?

From what I can see in your video, the opening explorer is only available as a standalone tool, separate from the analysis board. I think that it's a must that it's a part of the analysis board, so you can use it when analyzing games. 

I think designing the ui is the difficult part. The table itself takes up a lot of space, and you also need to be able to switch between the masters, lichess, and player db, and the different settings for each of them. All while still being available from within the analysis board.

One option that comes to mind is to be able to swipe between the most list and the opening explorer. It has the benefit of always being able to see the board, which I think would be a big plus, but it leaves less space for the explorer compared to having a whole screen dedicated to it.

Any and all ideas are appreciated, since I feel like it can be quite tricky to get it right.

veloce commented:
This one will wait in the queue for the review, until I've dealt with older PRs. 

Just a couple of quick notes from what I've read here:

- a separate dedicated screen is a must have indeed (I think accessed from the analysis screen with a bottom bar icon, and also separately from the tools tab)
- but it doesn't prevent having a split view in the analysis screen as you did (with a settings disabled by default I'd say); can be useful for devices where vertical space is not an issue
- for wins/draws/loss, width according to percentage is a must have
- we should try to reduce the vertical padding of table rows but we shouldn't shrink them too much (as in the 2nd video), since these rows are also tappable targets.

veloce commented:
Do you need a review or feedback about the UI @Mauritz8 ? Please send a screen recording if you need.

Mauritz8 commented:
Overall, I'm quite happy with how it has turned out, but feedback is, as always, welcome. Functionality wise, there are a few things still missing, but a few of them can probably be left for another pr:
* Add unit tests for api requests.
* The specific lichess player endpoint streams the response using newline delimited json, which I didn't account for. To support it I'll have to change the approach quite a bit I think.
* Cache results.
* Ability to view master games. Not sure how this should be done, I quickly looked at the website and it seems like it uses websockets for this, but I don't know why that would be necessary.
* Tablebase, should be separate pr.
* Some other small adjustments.

[Screencast from 2024-08-03 17:17:48.webm](https://github.com/user-attachments/assets/329bb37a-3326-488a-962d-a4071586ad89)

ijm8710 commented:
I usually have a ton of feedback. I have very little, a sign that this looks awesome. Great work dude.

Only thing is was the percentage of the move removed intentionally. To me I like that being there to help highlight the relative popularity of comparative moves.
I do prefer it columned like on mobile more so than the slash in your original recording above.

ijm8710 commented:
Oh and wouldn't make sense to get the wiki book in as well at some point

veloce commented:
This looks awesome! Excellent work. I'll review and give more detailed feedback later.

To view master games, I think they are imported in lichess so it should just be a game ID? Also newline delimited JSON is already supported in the app, so it should not be too hard.

Mauritz8 commented:
I spent a lot of time on this, so I'm glad you like it! 
> Only thing is was the percentage of the move removed intentionally. To me I like that being there to help highlight the relative popularity of comparative moves.
I do prefer it columned like on mobile more so than the slash in your original recording above.

Horizontal space was a bit of an issue, so I made it show up on long press instead. Having a separate column for it would take to much space, unless the column name would be very short. And like you, I didn't really like how it looked when separated with just a slash or space.

> Oh and would it make sense to get the wiki book in as well at some point

I could make the opening name below the board be a link to the wikibook page. If integrating it into a dedicated screen is a wish, it is probably best left for later.

ijm8710 commented:
> I spent a lot of time on this, so I'm glad you like it! 

❤️


> Horizontal space was a bit of an issue, so I made it show up on long press instead. Having a separate column for it would take to much space, unless the column name would be very short. And like you, I didn't really like how it looked when separated with just a slash or space.

Hmm this is one time I'll disagree with ya. I still think percentage. Is important enough to show outright rather than on a long press. What about scaling font size down? Is there any other way?


> I could make the opening name below the board be a link to the wikibook page. If integrating it into a dedicated screen is a wish, it is probably best left for later. 

This totally works for me. Only request is link opens in-app browser as opposed to safari redirect

veloce commented on lib/src/view/opening_explorer/opening_explorer_screen.dart:
```diff
@@ -0,0 +1,845 @@
+  @override
+  Widget build(BuildContext context, WidgetRef ref) {
+    final ctrlProvider = analysisControllerProvider(pgn, options);
+    final position = ref.watch(ctrlProvider.select((value) => value.position));
```
I think we should not use all the `select()` here. As explained in [riverpod doc](https://riverpod.dev/docs/advanced/select#filtering-widgetprovider-rebuild-using-select), it's not necessary here and increase complexity.

veloce commented on lib/src/view/opening_explorer/opening_explorer_screen.dart:
```diff
@@ -0,0 +1,845 @@
+    final isRootNode = ref.watch(
+      ctrlProvider.select((s) => s.currentNode.isRoot),
+    );
+    final nodeOpening =
```
remove all the `.select`, cf. previous comment.

veloce commented on lib/src/view/opening_explorer/opening_explorer_screen.dart:
```diff
@@ -0,0 +1,845 @@
+              )
+            else
+              Expanded(
+                child: SingleChildScrollView(
```
The `SingleChildScrollView` might cause performance issues but let's leave it for now.

veloce commented on lib/src/view/opening_explorer/opening_explorer_screen.dart:
```diff
@@ -0,0 +1,845 @@
+                        whiteWins: openingExplorer.white,
+                        draws: openingExplorer.draws,
+                        blackWins: openingExplorer.black,
+                        ctrlProvider: ctrlProvider,
```
As explained in [riverpod doc](https://riverpod.dev/docs/essentials/do_dont#prefer-refwatchreadlisten-and-similar-apis-with-statically-known-providers), statically known providers are preferred. So you should pass only the `analysisOptions` and get the provider from them.

veloce commented on lib/src/view/opening_explorer/opening_explorer_screen.dart:
```diff
@@ -0,0 +1,845 @@
+        child: CircularProgressIndicator(),
+      ),
+      error: (error, stackTrace) => Center(
+        child: Text(error.toString()),
```
A `debugPrint` would be nice here.

veloce commented on lib/src/view/opening_explorer/opening_explorer_screen.dart:
```diff
@@ -0,0 +1,845 @@
+                  onTap: () => ref
+                      .read(ctrlProvider.notifier)
+                      .onUserMove(Move.fromUci(move.uci)!),
+                  child: Container(
```
A `Container` that only has a `padding` should be a `Padding` widget.

veloce commented on lib/src/view/opening_explorer/opening_explorer_screen.dart:
```diff
@@ -0,0 +1,845 @@
+                  onTap: () => ref
+                      .read(ctrlProvider.notifier)
+                      .onUserMove(Move.fromUci(move.uci)!),
+                  child: Container(
```
Padding. I won't put the comment again but should be applied to all `Container`.

veloce commented on lib/src/view/opening_explorer/opening_explorer_screen.dart:
```diff
@@ -0,0 +1,845 @@
+            case OpeningDatabase.lichess:
+            case OpeningDatabase.player:
+              final gameId = GameId(game.id);
+              final archivedGame = await ref.read(
```
This is not ideal but let's leave it like that for now. The `ArchivedGameScreen` should take either a `GameId` or `LightGameData` and load the game if it is a game id, like it is done on other screens.

veloce commented on lib/src/view/opening_explorer/opening_explorer_screen.dart:
```diff
@@ -0,0 +1,845 @@
+      .read(analysisControllerProvider(pgn, options).notifier)
+      .userPrevious();
+
+  Future<void> _showAnalysisMenu(BuildContext context, WidgetRef ref) {
```
Should be renamed `_showExplorerMenu` perhaps? And also we could avoid code duplication with the analysis menu here I think.

veloce reviewed (commented):
This is really an impressive piece of work! Thank you very much.

Could you solve the conflicts, and also send another screen recording with a tablet please? (landscape as well).

I'm really excited to have this in the app!

Btw, it is still marked as draft. What is missing?

Mauritz8 commented:
Thanks❤️, I'm also really hyped about this! Since your review, I've added a couple more things:
* Support for the player endpoint, which streams the results while it indexes more games. So, now one can see the results before it has fully indexed, and it also shows a loading indicator so the user knows it's not fully done. 
* The opening explorer data is now cached for every position, to avoid uneccesary requests to the server.
* I made the opening name a link to it's wikibooks page.
* The percentage is now shown directly in the table, instead of on long press.
* Unit tests for parsing the api responses

I haven't addressed any of the things you mentioned in your review just yet. I'll also upload screen recordings later.

Mauritz8 commented on lib/src/view/analysis/analysis_screen.dart:
```diff
@@ -796,6 +624,19 @@ class _BottomBar extends ConsumerWidget {
+              Expanded(
+                child: BottomBarButton(
+                  label: context.l10n.openingExplorer,
+                  onTap: () => pushReplacementPlatformRoute(
```
I had made it so that you could go to the analysis board from the opening explorer as well. So users could go back and forth multiple times which would clutter the navigation stack, and it would take many clicks to exit afterwards. On second thought though, I've now removed the possibility to go from the explorer to the analysis board, which gets rid of the problem.

veloce replied:
Yeah I think going from analysis to explorer is natural and you want to have support for back button.

If you open the explorer from the tools tab, you're likely to want only the explorer feature.

Mauritz8 commented:
I've now addressed all of your comments. I've made some screen recordings, the emulator starts lagging quite a lot when I'm recording unfortunately.

[Screencast from 2024-08-09 13:11:28.webm](https://github.com/user-attachments/assets/c231f1e1-763b-4549-9085-186755ccbc54)

[Screencast from 2024-08-09 13:21:26.webm](https://github.com/user-attachments/assets/b3618459-9cdc-477f-9a9e-35308bca5609)

[Screencast from 2024-08-09 13:24:59.webm](https://github.com/user-attachments/assets/f3c86007-8a2d-40e2-b304-914490a099c8)

Mauritz8 commented:
close #456 
close #872

veloce reviewed (commented):
I must thank you again for your work, I know it's a lot.

This review is mostly about the cache implementation which I think should be changed.

Another thing that we must keep in mind are widget tests. This feature is important so it must be tested I think. For a start we'd need at least some basic tests: the screen loads properly (we can count the entries), for each database type; When that is done we can see to add more tests (click on a entry, settings, etc.)

veloce commented:
Thanks for the recordings! The UI is slick, and looks nice on both phones and tablets.

One suggestion: we could have a preloaded chip with the current user when you choose "player" database. Because most of the time users want to see their game.

Mauritz8 commented on lib/src/model/opening_explorer/opening_explorer_repository.dart:
```diff
@@ -0,0 +1,196 @@
+part 'opening_explorer_repository.freezed.dart';
+
+@riverpod
+Stream<OpeningExplorerEntry> openingExplorer(
```
Thanks for the feedback, I can definitely see now why my cache implementation is subpar in many ways. One thing though, you say that a `StreamProvider` can be replaced by an `AsyncNotifier`, but I'm not really sure how that would be done. They yield keyword only works in combination with stream types, and using a regular return statement wouldn't work since we need to actually stream the results.

veloce replied:
You definitely can convert a StreamProvider to an AsyncNotifier. I've done it already in the app. I found that StreamProvider could lead to some subtle problems hard to see first, that's why I converted them.

An AsyncNotifier will produce state changes over time, that you can `watch`. So it does what a stream can do, except it is not a stream. To produce these state changes from a Stream you'd not use `for` and `yield` but `listen` to the stream inside the provider implementation.

Mauritz8 replied:
I've converted it now.

Mauritz8 commented on lib/src/model/opening_explorer/opening_explorer_repository.dart:
```diff
@@ -0,0 +1,196 @@
+}
+
+@freezed
+class OpeningExplorerCacheKey with _$OpeningExplorerCacheKey {
```
Since this is used as a map key we actually need the hashcode and equals methods generated by freezed.

Mauritz8 commented on lib/src/model/analysis/analysis_controller.dart:
```diff
@@ -454,6 +457,16 @@ class AnalysisController extends _$AnalysisController {
     }
   }
 
+  String _wikiBooksUrl(UciPath path) {
```
Yes, usually only the first couple moves of a game will have a valid url. I'll add a move number cap.

Mauritz8 commented:
I've refactored the cache implementation, it's still exposing a stream though (see my previous comment). I also added a move number cap to try and fetch the wikibooks url, and now it selects the current user as default for the player db.

I added a few widget tests as well, that check that the data loads for each db type. I'm not sure what the preferred way to mock a provider is, when it needs different state between test cases. I did a bit of a hack to change the database type between tests, but there's probably a better way.

veloce commented on lib/src/view/opening_explorer/opening_explorer_screen.dart:
```diff
@@ -0,0 +1,845 @@
+
+  @override
+  Widget build(BuildContext context) {
+    final ctrlProvider =
```
Since this is assigned with the return of `ref.watch` the name is not good here. What we have is the state of analysisController. So `analysisState` would be the proper name.

Mauritz8 commented on lib/src/model/common/http.dart:
```diff
@@ -433,6 +433,43 @@ extension ClientExtension on Client {
+      return response.stream
+          .map(utf8.decode)
+          .where((e) => e.isNotEmpty && e != '\n')
+          .map((e) => jsonDecode(e) as Map<String, dynamic>)
```
I think we have to convert the bytes to a string first in order to do the `where`, but the other two can be combined:
```dart
return response.stream
    .map(utf8.decode)
    .where((e) => e.isNotEmpty && e != '\n')
    .map((e) {
      final json = jsonDecode(e) as Map<String, dynamic>;
      return mapper(json);
    });
```

veloce replied:
Yeah probably. I wonder if the `e != '\n'` is really necessary btw. I don't remember why I put that in the first place.

Mauritz8 replied:
Yes it's needed, as empty lines may be sent to avoid timeouts.

Mauritz8 commented:
I've updated everything now.

veloce commented:
I must thank you again for your work @Mauritz8 . It is really awesome to see the explorer in the app finally 🎉 

I've added only cosmetic changes to your PR, and improved the loading effect (it doesn't blink anymore).

ijm8710 commented:
Hey @Mauritz8 had two questions?

1) when clicking a popular opening in mobile web, I see some openings open a page like this (so not directly wikibooks but a more comprehensive native page; in fact this page then links to wikibooks). Should the app do similarly? ![image](https://github.com/user-attachments/assets/67b04d5c-9251-4d56-a27b-78732c350b1c)
2) I've noticed that if I use opening explorer to analyze an opening, then open up like a masters game, that subsequent masters game if I go to game review and then click the opening explorer from there, the title of the opening does not open wikibooks same way it would if I came directly from the opening explorer itself. It's kinda like opening an opening explorer from
An opening explorer. Should it? The opening explorer screens look the same so would expect them to operate the same regardless if you're in one inside another.

Mauritz8 commented:
1. Having a native wikibooks page would definitely provide better and more seamless experience. I might not have time to work on it anytime soon though, but feel free to open an issue and I might get around to working on it.
2. Unfortunately (or fortunately?) I'm unable to reproduce this, for me the link opens just fine either way.

ijm8710 commented:
> 1. Having a native wikibooks page would definitely provide better and more seamless experience. I might not have time to work on it anytime soon though, but feel free to open an issue and I might get around to working on it.

Sorry not necessarily more native. But what I mean is the page that opens first in mobile web is not the wikibooks page. There's a more dedicated page that opens first that has a link to wikibooks page. Basically wondering if that's the page we should first have access to. If you try on mobile web perhaps you'll see what page I mean . It's just a different link I was curious about.
> 
> 2. Unfortunately (or fortunately?) I'm unable to reproduce this, for me the link opens just fine either way.

I explained the repro a bit wrong. If you go to your open game history and click analysis and then click the explorer button, the opening name doesnt click. If you can't repro off that I'll include a video but I think this time you probably might be able to repro

EmmetSchuler commented:
@ijm8710 I think this will eventually be its own tool, instead of only linking to the opening explorer. Currently in Beta, it's Lichess's way of offering their own crowdsourced explanations to openings and their variations without relying on the (in my opinion) often vague wiki books explanations.
Because this feature is relatively new even for the website, unless someone really wants to make a PR for it I don't think it's too high on priority.

EmmetSchuler commented:
Right now I believe it uses wikibooks as a baseline and are slowly filling it in with original content

ijm8710 commented:
You're not following my question emmet. I'm not asking for it to be  more native. I'm saying currently it's just a web redirect but the page provided is different from mobile web

Mauritz8 commented:
I see, it should just link to the lichess.org page instead. I like that, and it seems like an easy thing to change.

A recording of the bug would be appreciated.

EmmetSchuler commented:
> I'm not asking for it to be  more native. I'm saying currently it's just a web redirect but the page provided is different from mobile web. Shouldn't we use the same one?

I see, sorry for the misunderstanding

ijm8710 commented:
> I see, it should just link to the lichess.org page instead. I like that, and it seems like an easy thing to change.
> 
> 
> 
> A recording of the bug would be appreciated.

Awesome glad you agree and here's video of other. When there's a pause it doesn't open linkhttps://github.com/user-attachments/assets/7044f89b-2ba9-49be-a3b2-9133e6023d6c

Mauritz8 commented:
This recording doesn't show what you described with opening master games. Do you have an example where it works for a certain position one time, but not after opening the explorer from a master game? If not, it seems like it's an issue with certain openings at all times, such as the moscow variation you showed.

ijm8710 commented:
Perhaps that's a better way of describing the issue. I probably interpreted it wrong. Either way Moscow does work in mobile and has a wiki book so would assume it should work.

ijm8710 commented:
Actually yes I have an example. If you go to the 8th master game from top (the ones that show up without  making a first move) and from analysis you jump right to c5?! It will not work. But if you move one move back it will work. But both have the exact same name I believe

Mauritz8 commented:
Wikibooks has a different page for each sequence of moves, so even if two positions have the same opening name in lichess they will have different wikibook pages. In this case, after c5 there is no wikibook page.

ijm8710 commented:
Ah okay that makes sense! Apologies for the long road to show you that.

On mobile web, since it opens that OTHER beta page I referenced rather than wikibooks directly, it opens up "Indian Defence with e6, Nf3"

If you were to use this OTHER beta page, would it be able to do it even in the app too off of c5?! As well or is there some extra magic they do to basically go off move name rather than the entire sequence if there's an innacurzcy to the sequence line opening?

Mauritz8 commented:
Linking to the lichess.org page, like the website does, should be quite easy. Then there will be a link even after the move c5.

Mauritz8 commented:
I created a pr now #989
